src/main/kotlin/com/atlassian/performance/tools/infrastructure/dataset/ObsoleteHttpDatasetPackage.kt

```
@@ -16,21 +11,13 @@ internal class ObsoleteHttpDatasetPackage(
     private val downloadTimeout: Duration
 ) : DatasetPackage {
 
-    private val datasetBucket = URI("https://s3.eu-central-1.amazonaws.com/jira-soke-tests-eu/")
-
     override fun download(
         ssh: SshConnection
     ): String {
-        val unzipCommand = FileArchiver().pipeUnzip(ssh)
-        TaskTimer.time("download") {
-            IdempotentAction("download dataset") {
-                ssh.execute("wget -qO - ${datasetBucket.resolve(downloadPath)} | $unzipCommand", downloadTimeout)
-            }.retry(5, ExponentialBackoff(Duration.ofSeconds(2)))
-        }
-        return unpackedPath!!
+        throw UnsupportedOperationException("Download of this HttpDatasetPackage (downloadPath='$downloadPath') could not be executed. Create HttpDatasetPackage instance using the two-parameters constructor rather than the deprecated three-parameters one.")
```

Would it be better to fail faster in constructor instead of download method? Would save a lot of time for someone running the code

---

To fail-fast is good, but bear in mind backwards compatibility. As I wrote in the description "the download() operation for such-way constructed HttpDatasetPackage would fail anyways due to missing file in the deprecated jira-soke-tests-eu bucket. Therefore, I'm replacing the download() method at ObsoleteHttpDatasetPackage with throw UnsupportedOperationException implementation."
If someone's code does only construct such HttpDatasetPackage, but never download it, nether the missing file in S3 nor this change should break such code. Well, it's strange to construct HttpDatasetPackage without downloading it, still it's possible.

---

@gregrow I love your empathy towards our lib consumers. Behavioral compatibility is something worth preserving. We even officially promise it in each JPT lib public API definition (see the [CHANGELOG.md compatibility](https://github.com/atlassian/infrastructure/blob/master/CHANGELOG.md) section):
> [behavioral compatibility](http://cr.openjdk.java.net/~darcy/OpenJdkDevGuide/OpenJdkDevelopersGuide.v0.777.html#behavioral_compatibility) with behavioral contracts expressed via Javadoc

We've wanted a reasonable middle ground for what we promise regarding behavior. It's not as free as "we can change anything". It's not as restrictive as "we'll preserve JVM behavior for any possible set of inputs and environments".
Now, we never promised (in JavaDoc) that the `HttpDatasetPackage` constructors does not throw. So technically we're not breaking compatibility. But we can use our empathy (again props to @gregrow) to think whether it should have been promised. In that case, we should make it explicit, so that the consumers feel they have more to rely on and for maintainers to have an easier decision to make.

PS. JavaDoc is not just JavaDoc comments. Class and method names are part of this too. The names promise a certain behavior. And **if** (and that's a big if) we want to promise something more wide-spread like "no constructors are supposed to throw", then we can put it in package/module level JavaDoc or even in the compatibility docs directly.

---

PPS. IMHO we can throw here, because consumers are informed how they can fix the problem through the exception message.